### ISSUE TITLE
Remove constructor restriction for classes with test methods

### ISSUE THREAD

**Issue by kilahm** [HackPack/HackUnit#30]
This could require an API change.  Some options:

1. Require the class implement a `Testable` or `Test` or `HasTests` interface which defines a factory method for creating an instance.
2. Disable the constructor while the tests are being run.  This would likely be confusing for users and seems to go against the static typing mentality.

**Comment by fredemmott** [comment]
What is the constructor restriction? How is it implemented? Bit of a tangent to the goal of this issue, but the current thing sounds like an area where __ConsistentConstruct should be considered

**Comment by kilahm** [comment]
The ultimate goal of this would to allow test methods to live in the same class as the implementation code. Ideally directly below the method being tested.  That's why my first suggestion for an interface name was Testable. 

It's a different way of organizing the test code, so I'm not sure everyone would like it. I like the idea though. It's consistent with my router and factory container. There is also precedent in another language. http://dlang.org/spec/unittest.html

**Comment by kilahm** [comment]
Here's my thoughts on being able to instantiate arbitrary classes that happen to have test methods:
1. If the constructor does not require any parameters, just instantiate it through the Reflection API as is done currently.
2. If the constructor has required parameters, require that the class implement an interface requiring it to have a factory method which will be used to create the instance which becomes the context while running the test methods.

My proposed interface is:

```php
<?hh // strict

namespace HackPack\HackUnit\Contract;

interface Testable
{
  public static function testableInstance() : this;
}
```
@fredemmott @belka-ew thoughts?

**Comment by belka-ew** [comment]
To 2:
Maybe use annotations for the constructors. I think whether an object may require several constructors, for example if different tests need to pass different arguments to the constructor. For example ```<<Setup>>``` without argument for default constructor and ```<<Setup(`methodName`)>>``` for the single tests.

```php
<?hh
<<Test>>
public function testSomething(Assert $assert) : void
{
  // Do some testing here!
  $assert->int(2)->not()->eq(3);
  $assert
    ->whenCalled(() ==> {throw new \Exception(‘bad error)})
    ->willThrowClassWithMessage(\Exception::class, ‘bad error’)
    ;
}

<<Setup(`testSomething`)>>
public static function testableInstance() : this
{
}

<<Setup>>
public static function defaultTestableInstance() : this
{
}
```

The names and arguments are just an example. it can be ```<<Construct>>``` or ```<<Setup(`Test`, `methodName`)>>``` or what ever.
And it may be a good thing to prefix the attributes to avoid  conflicts. maybe something like ```<<UnitSetup>>``` or ```<<UnitTest>>``` or ```<<HUSetup>>```..

**Comment by fredemmott** [comment]
How about:

```
<?hh // strict

interface TestableBase {
  // ...
}

<<__ConsistentConstruct>>
abstract class Testable implements TestableBase {
  public function __construct() {
  }
}

abstract class TestableWithCustomConstructor implements TestableBase {
  public static abstract function testableInstance(): TestableBase;
}
```

Went for classes instead of interfaces as otherwise every Testable must implement a constructor, even if it's a no-op.

**Comment by kilahm** [comment]
I was thinking that a class wouldn't need to implement the `Testable` interface if it doesn't require any special constructor parameters.  In the most basic usage, you just need to annotate a method with `<<Test>>` and it will just run.  No need to implement/extend anything if you don't need any customization.  The instances are created using the Reflection API anyway, so really its just a matter of telling HackUnit how to create an instance.

I like the idea of using the attributes.  I'm reminded of the `@dataProvider` attribute used by PHPUnit, where the value of the attribute is the method that generates the data.  We could do something similar for which factory method to use for a particular test.  If one is not specified and the constructor requires parameters, the current `MalformedSuite` event can still be emitted, and the tests in that suite would be skipped.

We could also allow factories to be labeled and their labels referenced by the test attributes:

```php
<?hh

class MyClass
{
  <<Provides('MyClassWithMockA')>>
  public static function withMockDependencyA() : this
  {
    return new self(new MockDependencyA());
  }

  <<Provides('MyClassWithMockB')>>
  public static function withMockDependencyB() : this
  {
    return new self(new MockDependencyB());
  }

  public function __construct(private Dependency $needed) {}

  <<Test('MyClassWithMockA')>>
  public function testWithMockA(Assert $assert) : void
  {
    // Tests here
  }

  <<Test('MyClassWithMockB')>>
  public function testWithMockB(Assert $assert) : void
  {
    // Tests here
  }
}
```

I went with `Provides` because it is the same attribute I use in my factory container project.

**Comment by kilahm** [comment]
On the naming scheme: I'm not sure we should be worried about using very generic terms.  If the method is supposed to be a unit test, then annotating it as `<<Test>>` seems appropriate.  This would still be the case if you were using a different testing framework.  I think the argument also holds for Setup and Teardown.

**Comment by kilahm** [comment]
Of course using paramerized attributes this way requires two constant string to be identical which the type checker cannot catch....

**Comment by kilahm** [comment]
there would also need to be a way to define a default factory.

**Comment by kilahm** [comment]
@belka-ew I just noticed another detail from your example:  you have the factory method specifying which test method it should be used for.  I think it should be the other way around.  One factory could be used by multiple tests, but each test can only use one factory.

**Comment by belka-ew** [comment]
> On the naming scheme: I'm not sure we should be worried about using very generic terms. If the method is supposed to be a unit test, then annotating it as <<Test>> seems appropriate. This would still be the case if you were using a different testing framework. I think the argument also holds for Setup and Teardown.

agreed.

> @belka-ew I just noticed another detail from your example: you have the factory method specifying which test method it should be used for. I think it should be the other way around. One factory could be used by multiple tests, but each test can only use one factory.

Yes, you're right.

> there would also need to be a way to define a default factory.

```@Provides``` (or better ```@Provider``` then) without a parameter?

> Of course using paramerized attributes this way requires two constant string to be identical which the type checker cannot catch....

I think it isn't bad. An exception with a clear error description can be thrown if something is wrong.

**Comment by kilahm** [comment]
I went with `<<SuiteProvider('name')>>` for the factory attribute.  Just released version 0.5.0 that implements this.  It required a pretty big rewrite of the test loading/parsing system.
